workflow-ui/backstack-android/src/main/java/com/squareup/workflow1/ui/backstack/ViewStateFrame.kt

```diff
@@ -1,19 +1,179 @@
 package com.squareup.workflow1.ui.backstack
 
+import android.os.Bundle
 import android.os.Parcel
 import android.os.Parcelable
 import android.os.Parcelable.Creator
 import android.util.SparseArray
+import android.view.View
+import androidx.annotation.VisibleForTesting
+import androidx.annotation.VisibleForTesting.PRIVATE
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.Lifecycle.Event
+import androidx.lifecycle.Lifecycle.Event.ON_DESTROY
+import androidx.lifecycle.LifecycleEventObserver
+import androidx.lifecycle.LifecycleOwner
+import androidx.lifecycle.ViewTreeLifecycleOwner
+import androidx.savedstate.SavedStateRegistry
+import androidx.savedstate.SavedStateRegistryController
+import androidx.savedstate.SavedStateRegistryOwner
+import androidx.savedstate.ViewTreeSavedStateRegistryOwner
+import com.squareup.workflow1.ui.WorkflowLifecycleOwner
+import com.squareup.workflow1.ui.WorkflowUiExperimentalApi
+import com.squareup.workflow1.ui.backstack.ViewStateFrame.AttachState.Attached
+import com.squareup.workflow1.ui.backstack.ViewStateFrame.AttachState.Detached
 
 /**
  * Used by [ViewStateCache] to record the [viewState] data for the view identified
  * by [key], which is expected to match the `toString()` of a
  * [com.squareup.workflow1.ui.Compatible.compatibilityKey].
+ *
+ * This class manages both the Android view state traditionally saved and restored through the
+ * [View.onSaveInstanceState]/[View.onRestoreInstanceState], as well as the modern
+ * [SavedStateRegistry] for the view exposed via [ViewTreeSavedStateRegistryOwner].
+ *
+ * A frame can either be [Detached] if it's simply holding the saved data for a hidden view, or
+ * [Attached] when it's managing a visible view.
  */
-internal data class ViewStateFrame(
+@OptIn(WorkflowUiExperimentalApi::class)
+internal class ViewStateFrame(
```

Comment:
This class shares even more similarities now with things like `Fragment` and [`NavBackStackEntry`](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:navigation/navigation-runtime/src/main/java/androidx/navigation/NavBackStackEntry.java;l=60?q=backstackentry&sq=).